File: examples/Perl/taskwork2.pl
# Task worker - design 2 in Perl
# Adds pub-sub flow to receive and respond to kill signal

use strict;
use warnings;
use v5.10;

$| = 1; # autoflush stdout after each print

use ZMQ::FFI;
use ZMQ::FFI::Constants qw(ZMQ_PULL ZMQ_PUSH ZMQ_SUB);

use Time::HiRes qw(usleep);
use AnyEvent;
use EV;

# Socket to receive messages on
my $context = ZMQ::FFI->new();
my $receiver = $context->socket(ZMQ_PULL);
$receiver->connect('tcp://localhost:5557');

# Socket to send messages to
my $sender = $context->socket(ZMQ_PUSH);
$sender->connect('tcp://localhost:5558');

# Socket for control input
my $controller = $context->socket(ZMQ_SUB);
$controller->connect('tcp://localhost:5559');
$controller->subscribe('');

# Process messages from either socket

my $receiver_poller = AE::io $receiver->get_fd, 0, sub {
    while ($receiver->has_pollin) {
        my $string = $receiver->recv();

        print "$string.";    # Show progress
        usleep $string*1000; # Do the work
        $sender->send('');   # Send results to sink
    }
};

# Any controller command acts as 'KILL'
my $controller_poller = AE::io $controller->get_fd, 0, sub {
    if ($controller->has_pollin) {
        EV::break; # Exit loop
    }
};

EV::run;

File: examples/OCaml/taskwork2.ml
(**
 * Task worker - design 2
 * Adds pub-sub flow to receive and respond to kill signal
 *)
open Zmq
open Helpers

let () =
  with_context @@ fun ctx ->
  (* Socket to receive messages on *)
  with_socket ctx Socket.pull @@ fun receiver ->
  Socket.connect receiver "tcp://localhost:5557";

  (* Socket to send messages to *)
  with_socket ctx Socket.push @@ fun sender ->
  Socket.connect sender "tcp://localhost:5558";

  (* Socket for control input *)
  with_socket ctx Socket.sub @@ fun controller ->
  Socket.connect controller "tcp://localhost:5559";
  Socket.subscribe controller "";

  let items = Poll.mask_of [| (receiver, Poll.In); (controller, Poll.In) |] in

  (* Process tasks from either socket *)
  while true do
    let pollResults = Poll.poll items in
    let receiverEvents = pollResults.(0) in
    let ctrlEvents = pollResults.(1) in

    match receiverEvents with
    | Some _ ->
      let s = Socket.recv receiver in
      printfn "%S." s; (* Show progress *)
      sleep_ms (int_of_string s); (* Do the work *)
      Socket.send sender ""; (* Send results to sink *)
    | _ -> ();

    (* Any waiting controller command acts as 'KILL' *)
    match ctrlEvents with
    | Some _ -> exit 0; (* Exit program *)
    | _ -> ();
  done
